### ISSUE TITLE
VSExtension support VS 2019

### ISSUE THREAD

**Issue by chrpai** [wixtoolset/issues#5923]
Please provide answers to the following questions to help us narrow down, reproduce, and fix the problem. Fill out one section and delete the others

## Feature requests

VSExtension will need updating  to support detecting VS2019 properties.

**Comment by chrpai** [comment]
The c++ side is a bit of a stretch for me but I'm willing to work this.   Worst case I have to return it back to unassigned.   I've already started the  XML side of it.

**Comment by chrpai** [comment]
I sent an RFC email to the group describing my approach.   I was able to successfully make a private build for myself to handle my particular needs  so this can be accepted or deferred  doesn't really matter to me at this point.   Personally I'd like to see this in v3 at some point.

**Comment by chrpai** [comment]
Have code at:
https://github.com/chrpai/wix3/tree/vsext

**Comment by creage** [comment]
What about bugs on this new 2019 extension? Should we report them as usual?

I'm getting this one
```
Setup.Installer\setup.wixproj : error  : A numeric comparison was attempted on "$(MSBuildToolsVersion)" that evaluates to "Current" instead of a number, in condition " '$(MSBuildToolsVersion)' == '' OR '$(MSBuildToolsVersion)' < '4.0' ".  C:\Program Files (x86)\MSBuild\Microsoft\WiX\v3.x\wix.targets
```

So I had to manually alter my ``wix.targets`` file, changing failing lines to this

```xml
 <WixVersionTargetsPath Condition=" '$(MSBuildToolsVersion)' == '' OR '$(MSBuildToolsVersion)' == 'Current' OR '$(MSBuildToolsVersion)' &lt; '4.0'">wix200x.targets</WixVersionTargetsPath>
```

**Comment by BMurri** [comment]
This one's a known issue. Msbuild has promised a fix. A workaround is in this pull request: https://github.com/wixtoolset/VisualStudioExtension/pull/11

Get Outlook for Android<https://aka.ms/ghei36>

**Comment by chrpai** [comment]
I hoped the preview build on wixtoolset would have contained that PR but a decision seems to have been made to not include it yet likely in hopes MSBuild will fix their regression.   I'm guessing that by the time VS is released they will have fixed MSBuild of the extension will contain the PR.

In the meantime, I published a private build of VSIX  containing that pull request available here:

https://github.com/iswix-llc/iswix/releases/tag/4.12.6

**Comment by chrpai** [comment]
BTW that issue and this issue are different.  This issue is for the VSExtension to support detect VS2019.   That has not been released yet but is available from the link above via a private release of wix 3.14.

**Comment by chrpai** [comment]
Goodnews bad news....

The MSBuild issue is confirmed fixed in VS2019 Preview 4.

The bad news....  the work done under this  issue may now be broken under preview 4.  While I haven't tested a  v4 build,  this same logic had been ported to a private fork of WiX v3.14 and used to build IsWiX.  IsWiX uses these properties to register with VS2019.     It was all working up until today.  In VS2019 preview 4 the properties don't resolve and the IsWiX  addin is not registered.

**Comment by chrpai** [comment]
MSI (c) (60:94) [15:03:08:148]: Doing action: VSFindInstances
Action 15:03:08: VSFindInstances. 
Action start 15:03:08: VSFindInstances.
MSI (c) (60:D0) [15:03:08:164]: Invoking remote custom action. DLL: C:\Users\cxp5196\AppData\Local\Temp\1\MSI4F7B.tmp, Entrypoint: FindInstances
MSI (c) (60:F4) [15:03:08:166]: Cloaking enabled.
MSI (c) (60:F4) [15:03:08:166]: Attempting to enable all disabled privileges before calling Install on Server
MSI (c) (60:F4) [15:03:08:166]: Connected to service for CA interface.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_ROOT_FOLDER property. Its value is 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional'.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_IDE_VWD_PROJECTSYSTEM_INSTALLED property. Its value is '1'.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_IDE_MODELING_PROJECTSYSTEM_INSTALLED property. Its value is '1'.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_IDE_VB_PROJECTSYSTEM_INSTALLED property. Its value is '1'.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_IDE_VCSHARP_PROJECTSYSTEM_INSTALLED property. Its value is '1'.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_IDE_FSHARP_PROJECTSYSTEM_INSTALLED property. Its value is '1'.
MSI (c) (60!DC) [15:03:08:248]: PROPERTY CHANGE: Adding VS2017_IDE_VC_PROJECTSYSTEM_INSTALLED property. Its value is '1'.
Action ended 15:03:08: VSFindInstances. Return value 1.

**Comment by chrpai** [comment]
I was unable to repro this on a machine at home.  I will look at it at work again tomorrow.

**Comment by chrpai** [comment]
I've experienced this on another machine now.  It's almost as if machines that had previous versions of VS 2019  and upgraded to the latest work and new machines that go straight to 2019 latest don't work.

**Comment by chrpai** [comment]
I've now seen it work with a clean install of 2019 latest and I've seen a clean install of 2017 and 2019 come back as detecting 2019 but not 2017.   the sample project Setup.Configuration.CS.exe  detects both in all scenarios so personally I'm suspecting that my 2019 work done in C++ is incomplete and that it suffers some 2017/2019  interaction side effect.  Not suprising since I'm not really a c++ developer.  I'd reccomend considering this pull request suspect and have someone who knows c++  take a look at it.

**Comment by MichaelKetting** [comment]
Am I correct in assuming this is the thread, where support for VS 2019 for the WixToolset Visual Studio Extension is being requested?

If so, yes, here's hoping for an updated release now that VS 2019 is RTM. I would hate to not be able to upgrade my solution just because WixToolset is not yet support.

Best regards, Michael

**Comment by robmen** [comment]
Preview support for VS 2019 is available now here: http://wixtoolset.org/releases/.

It will be promoted to the VS Marketplace soon.

**Comment by RobHarrisAZ** [comment]
What is the status of this? The preview addon only seems to detect Community edition.
4/5/2019 10:35:05 AM - 	Identifier         : WixToolset.VisualStudioExtension.Dev16
4/5/2019 10:35:05 AM - 	Name               : Wix Toolset Visual Studio 2019 Extension
4/5/2019 10:35:05 AM - 	Author             : .NET Foundation and contributors
4/5/2019 10:35:05 AM - 	Version            : 0.9.29.61328
4/5/2019 10:35:05 AM - 	Description        : Visual Studio integration for the WiX Toolset - the most powerful set of tools available to create your Windows installation experience.
4/5/2019 10:35:05 AM - 	Locale             : en-US
4/5/2019 10:35:05 AM - 	MoreInfoURL        : http://wixtoolset.org/
4/5/2019 10:35:05 AM - 	InstalledByMSI     : False
4/5/2019 10:35:05 AM - 	SupportedFrameworkVersionRange : [4.0,)
4/5/2019 10:35:05 AM - 
4/5/2019 10:35:05 AM - 	SignatureState     : Unsigned
4/5/2019 10:35:05 AM - 	Supported Products : 
4/5/2019 10:35:05 AM - 		Microsoft.VisualStudio.Community
4/5/2019 10:35:05 AM - 			Version : [16.0,17.0)
4/5/2019 10:35:05 AM - 

I have VS 2019 Enterprise and it won't install.

**Comment by chrpai** [comment]
I use VS2019 Professional at work and it installs just fine.  I'll have to see if my VS subscription has Enterprise. I'll throw it on a VM if it does.

**Comment by robmen** [comment]
Votive works in VS2019 Enterprise. I've verified it. However, as I understand it, this issue is about updating the VisualStudio.wixext to support 2019 (it is not about Votive). We should stop discussing Votive on this issue. :smiley:

**Comment by RobHarrisAZ** [comment]
> 
> 
> I use VS2019 Professional at work and it installs just fine. I'll have to see if my VS subscription has Enterprise. I'll throw it on a VM if it does.
> 
> Did you use WiX with VS2017 Enterprise?

Yes. It works great on 2017 Enterprise but I'd like to get back to having 1 copy of VS installed and right now wixproj files are the only thing preventing that for me. I installed the 3.14 toolset but that didn't add the project types to VS 2019. For that as I understand it, I need the extension.

**Comment by chrpai** [comment]
Rob is correct.  This is the right issue.   You want issue  5915 which tracked creating VS2019 votive support.

That said,  I just threw VS2019 Enterprise, WiX 3.11.1 and the VS 2019 Votive extension on a VM and it all installed no issues for me. 

![image](https://user-images.githubusercontent.com/43178814/55648052-145e2000-5794-11e9-942c-5fd32f46ba49.png)

**Comment by BaconTentacles** [comment]
I can confirm that the extension preview does install just fine with VS Professional 2019 16.0.3.  Huzzah.

That being said, there is a nasty little bug with VS itself, so I gotta stick with VS 2017 for the time being.  Whoops.

https://developercommunity.visualstudio.com/content/problem/519499/could-not-copy-the-file-objx64debugmanagementexe-b.html

But, hey, it's not Wix's problem.  Yay?

**Comment by kdawg1406** [comment]
Does anyone know how to get the VS 2019 property for the root folder?

In VS 2015 I used  VS2015_ROOT_FOLDER

When I try to use VS2019_ROOT_FOLDER I get an error.

Greatly appreciate any assistance in figuring this out.

Karl

**Comment by BMurri** [comment]
My understanding is that VS removed the "root folder" for vs 2017/2019/future because they now allow SxS installations of any given VS, even if the same version.

You have to query the installed instances and deal with each one, now.

**Comment by davidnmbond** [comment]
Broken again with VS 2019 16.1.0

**Comment by chrpai** [comment]
What version of WiX are you using? 3.11 or the 3.14?  (You need the latter.)

**Comment by davidnmbond** [comment]
I'm using the version listed here: https://wixtoolset.org/releases/ as "Release Candidate of WiX Toolset Visual Studio 2019 Extension", though there is no indication of version number.
I have recorded a bug here: https://github.com/wixtoolset/issues/issues/6017
Note that there seems to be a workaround - just right click on the "incompatible" project and reload.

**Comment by MikeRosoft** [comment]
Right now it doesn't work in Visual Studio 2019 at all - the wixproj project doesn't open (it is reported as incompatible - "the application is not installed"). The 'reload project' option is disabled. I am using Visual Studio 16.4.3, extension 1.0.0.4, Wix 3.14.0.3316.

**Comment by MikeRosoft** [comment]
Update: After restarting the operating system, the 'reload' option became enabled.

**Comment by rseanhall** [comment]
v4 PR was accepted - https://github.com/wixtoolset/VisualStudio.wixext/pull/3
v3 PR was not since this issue wasn't approved for 3.x - https://github.com/wixtoolset/wix3/pull/478

**Comment by rseanhall** [comment]
I see that this actually was implemented in v3.14 in https://github.com/wixtoolset/issues/issues/5979.
